yule/Main.hs

```diff
@@ -21,45 +19,66 @@ main = do
     -- print options
     let filename = Options.input options
     src <- readFile filename
-    let (Object name code inners) = parseObject filename src
+    let inputObject = parseObject filename src
     let oCompress = Options.compress options
     when oCompress $ do
         putStrLn "Compressing sums"
-    let source = if oCompress then compress code else code
-    generatedYul <- runTM options (translateStmts source)
+    let compObject = if oCompress then compress inputObject
+                                  else inputObject
+    -- Yul "preobject" - lacking deployment code
+    yulPreobject@(YulObject yulName yulCode _) <- runTM options (translateObject compObject)
     let withDeployment = not (Options.runOnce options)
     let doc = if Options.wrap options
-        then wrapInSol (Name name) generatedYul
-        else wrapInObject (Name name) withDeployment generatedYul
+        then wrapInSol (Name yulName) (ycStmts yulCode)
+        else wrapInObject withDeployment yulPreobject
+
     putStrLn ("writing output to " ++ Options.output options)
     writeFile (Options.output options) (render doc)
 
 -- wrap in a Yul object with the given name
-wrapInObject :: Name -> Bool -> [YulStmt] -> Doc
-wrapInObject name deploy yul
-  | deploy    = ppr nested
-  | otherwise = ppr runtime
-  where
-    nested = YulObject "Deployable" deploycode [InnerObject runtime]
-    runtime = YulObject (show name) (YulCode stmts) []
-    cname = yulString (show name)
-    stmts = yul ++ retcode
-    retcode =
+wrapInObject :: Bool -> YulObject -> Doc
+wrapInObject deploy yulo@(YulObject name code inners)
+  | deploy    = ppr (createDeployment yulo)
+  | otherwise = ppr (YulObject name (addRetCode code) inners)
+
+addRetCode :: YulCode -> YulCode
+addRetCode c = c <> retCode where
+    retCode = YulCode
       [ calls "mstore" [yulInt 0, YIdent "_mainresult"]
       , calls "return" [yulInt 0, yulInt 32]
       ]
-    deploycode = YulCode
-      [ calls "mstore" [yulInt 64, YCall "memoryguard" [yulInt 128]]
-      , ylva "memPtr" (YCall "mload" [yulInt 64])
-      -- call constructor here
-      , calls "datacopy" [yulInt 0, dataoffset, datasize]
-      , calls "return" [yulInt 0, datasize]
-      ]
+    calls f args = YExp (YCall f args)
+
+deployCode :: String -> Bool -> YulCode
```

Comment:
I'm wondering if this needs to be done during codegen or if it can also be a desugaring pass?

Reply:
Yes, this is in the works, but that's what I start with.

In the upcoming PR, `deployCode` becomes much simpler:

```
deployCode _name withStart = YulCode $ go withStart where
  go True = [ [yulStmt| usr$start() |] ]
  go False = []
```